drivers/counter/counter_nxp_s32_sys_timer.c

```diff
@@ -0,0 +1,267 @@
+	.get_value = nxp_s32_sys_timer_get_value,
+	.set_alarm = nxp_s32_sys_timer_set_alarm,
+	.cancel_alarm = nxp_s32_sys_timer_cancel_alarm,
+	.set_top_value = nxp_s32_sys_timer_set_top_value,
+	.get_pending_int = nxp_s32_sys_timer_get_pending_int,
+	.get_top_value = nxp_s32_sys_timer_get_top_value,
+	.get_freq = nxp_s32_sys_timer_get_frequency
+};
+
+#define SYS_TIMER_CHANNEL_CFG(i, n)				\
+	{							\
+		.hwChannel = i,					\
+		.callback = &nxp_s32_sys_timer_##n##_callback,	\
+		.callbackParam = i,				\
+		.channelMode = STM_IP_CH_MODE_ONESHOT,		\
+	}
+
+#define SYS_TIMER_ISR_DECLARE(n)	\
+	extern void SYS_TIMER_ISR(CONFIG_NXP_S32_RTU_INDEX, n)(void)
+
+#define SYS_TIMER_INIT_DEVICE(n)							\
+	SYS_TIMER_ISR_DECLARE(n);							\
+											\
+	void nxp_s32_sys_timer_##n##_callback(uint8_t chan_id)				\
+	{										\
+		const struct device *dev = DEVICE_DT_GET(SYS_TIMER_NODE(n));		\
+		const struct nxp_s32_sys_timer_config *config = dev->config;		\
+		struct nxp_s32_sys_timer_data *data = dev->data;			\
+		struct nxp_s32_sys_timer_chan_data *ch_data = &data->ch_data[chan_id];	\
+		counter_alarm_callback_t cb = ch_data->callback;			\
+		uint32_t val;								\
+											\
+		ch_data->callback = NULL;						\
+		if (cb) {								\
+			val = Stm_Ip_GetCounterValue(config->instance);			\
+			cb(dev, chan_id, val, ch_data->user_data);			\
+		}									\
+	}										\
+											\
+	static int nxp_s32_sys_timer_##n##_init(const struct device *dev)		\
+	{										\
+		IRQ_CONNECT(DT_IRQN(SYS_TIMER_NODE(n)),					\
+			    DT_IRQ(SYS_TIMER_NODE(n), priority),			\
+			    SYS_TIMER_ISR(CONFIG_NXP_S32_RTU_INDEX, n),			\
+			    DEVICE_DT_GET(SYS_TIMER_NODE(n)),				\
+			    DT_IRQ(SYS_TIMER_NODE(n), flags));				\
+		irq_enable(DT_IRQN(SYS_TIMER_NODE(n)));					\
+											\
+		return nxp_s32_sys_timer_init(dev);					\
+	}										\
+											\
+	static struct nxp_s32_sys_timer_data nxp_s32_sys_timer_data_##n;		\
+											\
+	static const struct nxp_s32_sys_timer_config nxp_s32_sys_timer_config_##n = {	\
+		.info = {								\
+			.max_top_value = SYS_TIMER_MAX_VALUE,				\
+			.freq = (DT_PROP(SYS_TIMER_NODE(n), clock_frequency)),		\
+			.channels = SYS_TIMER_NUM_CHANNELS,				\
+			.flags = COUNTER_CONFIG_INFO_COUNT_UP,				\
+		},									\
+		.hw_cfg = {								\
+			.stopInDebugMode = DT_PROP(SYS_TIMER_NODE(n), freeze),		\
+			.clockPrescaler = DT_PROP(SYS_TIMER_NODE(n), prescaler) - 1,	\
+		},									\
+		.ch_cfg = {								\
+			LISTIFY(SYS_TIMER_NUM_CHANNELS, SYS_TIMER_CHANNEL_CFG, (,), n)	\
+		},									\
+		.instance = SYS_TIMER_INSTANCE_ID(n),					\
+	};										\
+											\
+	DEVICE_DT_DEFINE(SYS_TIMER_NODE(n),						\
+			 nxp_s32_sys_timer_##n##_init,					\
+			 NULL,								\
+			 &nxp_s32_sys_timer_data_##n,					\
+			 &nxp_s32_sys_timer_config_##n,					\
+			 POST_KERNEL,							\
+			 CONFIG_COUNTER_INIT_PRIORITY,					\
+			 &nxp_s32_sys_timer_driver_api);
+
+#if DT_NODE_HAS_STATUS(SYS_TIMER_NODE(0), okay)
```

Comment:
Can these initialization macros be replaced with [DT_INST_FOREACH_STATUS_OKAY](https://docs.zephyrproject.org/latest/build/dts/api/api.html#c.DT_INST_FOREACH_STATUS_OKAY)?

Reply:
The HAL for NXP S32 is index based and in general the interrupt handlers are defined per instance, so the actual hw instance index is still needed. I will address this for all existing drivers so far in another pr, to follow a more general approach. Changes in the baremetal driver will be required in some cases.